### PR TITLE
KAFKA-5474: Streams StandbyTask should no checkpoint on commit if EOS is enabled

### PR DESCRIPTION
**Pull request by mjsax**
<strike> - actual fix for `StandbyTask#commit()` </strike>

Additionally (for debugging):
 - EOS test, does not report "expected" value correctly
 - add `IntegerDecoder` (to be use with `kafka.tools.DumpLogSegments`)
 - add test for `StreamTask` to not checkpoint on commit if EOS enabled

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @enothereska @bbejeck @dguy @guozhangwang @ijuma 

This is a blocker for 0.11.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5467/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5482/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by mjsax** [comment]
System test passed. I'll do some more reruns to make sure it's fixed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/343/

New run with 20 repeats: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/344/

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java [hunk 1]
```diff
@@ -92,7 +92,9 @@ public void resume() {
     public void commit() {
         log.trace("{} Committing", logPrefix);
         stateMgr.flush();
-        stateMgr.checkpoint(Collections.<TopicPartition, Long>emptyMap());
+        if (!eosEnabled) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java]: While reading the code: when we call `StandbyTask.close()` it calls `commit()` but not `suspend()`; with this change when EOS is turned on we will not write checkpoint files when closing the standby task, I think this is not right?

File: core/src/main/scala/kafka/serializer/Decoder.scala [hunk 2]
```diff
@@ -54,17 +54,19 @@ class StringDecoder(props: VerifiableProperties = null) extends Decoder[String]
 }
 
 /**
-  * The string decoder translates bytes into strings. It uses UTF8 by default but takes
-  * an optional property serializer.encoding to control this.
+  * The long decoder translates bytes into longs.
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/serializer/Decoder.scala:57]: Good catch! Could we add the corresponding encoder in `Encoder` as well?

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 3]
```diff
@@ -334,18 +334,18 @@ private static void verifyMin(final Map<TopicPartition, List<ConsumerRecord<byte
...
                     final int receivedValue = integerDeserializer.deserialize(receivedRecord.topic(), receivedRecord.value());
                     final String key = stringDeserializer.deserialize(input.topic(), input.key());
-                    final Integer value = integerDeserializer.deserialize(input.topic(), input.value());
+                    final int value = integerDeserializer.deserialize(input.topic(), input.value());
 
 
                     Integer min = currentMinPerKey.get(key);
                     if (min == null) {
-                        min = value;
+                        min = Integer.MAX_VALUE;
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: nit: 

```
if (min == null)
    min = value;
else
    min = Math.min(...)
```

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 4]
```diff
@@ -383,8 +383,8 @@ private static void verifySum(final Map<TopicPartition, List<ConsumerRecord<byte
                     sum += value;
                     currentSumPerKey.put(key, sum);
 
-                    if (!receivedKey.equals(key) || receivedValue != sum) {
-                        throw new RuntimeException("Result verification failed for " + receivedRecord + " expected <" + key + "," + value + "> but was <" + receivedKey + "," + receivedValue + ">");
+                    if (!receivedKey.equals(key) || receivedValue != sum.longValue()) {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: Ditto above: `if (sum == null) sum = value else sum += value`.

**Review by guozhangwang — COMMENTED** [review]
Left some comments.

**Comment by ijuma** [comment]
Can't we add any unit or integration tests for this bug fix?

**Review by dguy — COMMENTED** [review]
@mjsax why should the standby task not checkpoint on commit when EOS is enabled?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java]: We write the checkpoint file when closing the `StateManager` -- so this is fine. Nevertheless, I did double check the current behavior and made some minor changes. Also updated https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Streams+Architecture

The code should align to this design (please also verify the design).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5525/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5511/
Test PASSed (JDK 8 and Scala 2.12).

**Review by bbejeck — APPROVED** [review]

**Comment by mjsax** [comment]
@dguy Thanks for you comment. I made me think about this once more. Digger deeper, we discovered that only `trunk` is broken (cf. https://github.com/mjsax/kafka/commit/90b2a2bf664e4e40d4cd1b46c72732c5edb97cf9#commitcomment-22667101) Thus, there is no reason to disable checkpointing.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5548/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5534/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by guozhangwang** [comment]
The rest of the code clean-up / minor fixes LGTM. Merged to trunk and 0.11.0.

**Pull request closed without merging**